### PR TITLE
Percent-encode invalid characters with request target

### LINKED ISSUE #1585: parse_url no longer accepts [] in the URL
With urllib3 1.25.1:

```
>>> parse_url('https://example.com/somevar[]')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/testrunner/testvenv/lib/python2.7/site-packages/urllib3/util/url.py", line 212, in parse_url
    six.raise_from(LocationParseError(url), None)
  File "/home/testrunner/testvenv/lib/python2.7/site-packages/urllib3/packages/six.py", line 718, in raise_from
    raise value
urllib3.exceptions.LocationParseError: Failed to parse: https://example.com/somevar[]
```

Using 1.24.2:

```
>>> parse_url('https://example.com/somevar[]')
Url(scheme='https', auth=None, host='example.com', port=None, path='/somevar[]', query=None, fragment=None)
```

```
>>> parse_url('https://example.com/?somevar["whatever"]')
Url(scheme='https', auth=None, host='example.com', port=None, path='/', query='somevar["whatever"]', fragment=None)
```

This is used by various bits of software (ElasticSearch for instance):

```
>>> parse_url('http://example.com/api/search/?grouping=datasets&q=dontcare&projection=["variables.nonexistingfield"]')
Url(scheme='http', auth=None, host='example.com', port=None, path='/api/search/', query='grouping=datasets&q=dontcare&projection=["variables.nonexistingfield"]', fragment=None)
```

### PR DESCRIPTION
**Pull request by sethmlarson**
Closes #1585. cc @theacodes @sigmavirus24

### CODE REVIEW

**Comment by sethmlarson** [comment]
Yeah I need to add some tests so we don't regress on this behavior.

**Comment by sethmlarson** [comment]
@theacodes This is ready for review I believe.

**Comment by sethmlarson** [comment]
I thought about putting the `_encode_invalid_chars` calls within the conditional for normalizable schemes but `http+unix` isn't impacted by changing the path or query so I think we're safe to unconditionally do that operation?

**Comment by sethmlarson** [comment]
So it's going to sound strange but I can't think of any scenarios that would get us an exception out of the `validator.check_validity_of()` branch? We already percent-encode all characters that aren't valid within path, query, and fragment and the authority (userinfo, host, and port) are handled in the `uri_ref.normalize()`. Might be a good job for hypothesis to find some good test cases if there even are any?

File: src/urllib3/util/url.py [hunk 1]
```diff
@@ -252,7 +251,7 @@ def idna_encode(name):
             *validator.COMPONENT_NAMES
         ).validate(uri_ref)
     except ValidationError:
-        six.raise_from(LocationParseError(url), None)
+        return six.raise_from(LocationParseError(url), None)
```
Comment (sigmavirus24) [inline src/urllib3/util/url.py]: Is the `return` here actually necessary?

Comment (sethmlarson) [reply src/urllib3/util/url.py]: Not at all, however IDEs don't recognize that `six.raise_from()` won't allow execution to propagate. Should I remove it? Don't mind either way.

**Comment by pquentin** [comment]
I was wondering, is it expected that util/url.py is not at 100% coverage?

**Comment by sethmlarson** [comment]
Yeah, I need to figure out a scenario and investigate the failure on 2.7.

**Comment by sethmlarson** [comment]
@pquentin And by "yeah" I mean "no", my Minnesotan is leaking through. :)

**Pull request merged**